src/api/auth/register/register.controller.spec.ts

```diff
@@ -0,0 +1,63 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { UserService } from '../../user/user.service';
+import { RegisterUserDto } from './dto/register-user.dto';
+import { RegisterController } from './register.controller';
+import { RegisterService } from './register.service';
+import { createMock } from '@golevelup/ts-jest';
+
+describe('RegisterController', () => {
+  let controller: RegisterController;
+  let registerService: RegisterService;
+
+  beforeEach(async () => {
+    const RegisterServiceProvider = {
+      provide: RegisterService,
+      useFactory: () => ({
+        register: jest.fn(),
+      }),
+    };
+
+    const UserServiceProvider = {
+      provide: UserService,
+      useFactory: () => ({
+        findAll: jest.fn(),
+        getUser: jest.fn(),
+        createUser: jest.fn(),
+      }),
+    };
+
+    const module: TestingModule = await Test.createTestingModule({
+      controllers: [RegisterController],
+      providers: [RegisterServiceProvider, UserServiceProvider],
+    }).compile();
+
+    controller = module.get<RegisterController>(RegisterController);
+    registerService = module.get<RegisterService>(RegisterService);
+  });
+
+  it('should be defined', () => {
+    expect(controller).toBeDefined();
+  });
+
+  describe('register', () => {
+    it('calls service.register() method', async () => {
```

Comment:
It seems like the validation error checking is all done in the actual API response code, which is a layer of abstraction removed from where it should be happening. You should be testing validation (which includes a success and failures for all validations [e.g., blank username, blank password, password too short, password too long, password without number, password without special character, etc..]) at the place where the code actually runs the validation, not the API response

Reply:
@kaitwalla When you refer to API response code? Are you referring to Nest's built-in response code?

Reply:
The API response code I was referring to was the e2e tests, which appears to be on the only place where the validation logic is being checked. In an ideal situation, the logic would be checked where it's being implemented. That's slightly more difficult to do in this scenario because so much of it is being handled by library code, but the tests should be at the closest point to the logic (in this case, from what I can ascertain, at the service level where you're actually calling .register)

Reply:
👍